include/triton/core/tritonbackend.h

```diff
@@ -983,6 +983,15 @@ TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_ModelRepository(
     TRITONBACKEND_Model* model, TRITONBACKEND_ArtifactType* artifact_type,
     const char** location);
 
+/// Get the location of the model (moedl.py), provided by a python based
+/// backend.
+///
+/// \param model The model.
+/// \param location Returns the location.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error* TRITONBACKEND_BackendModelLocation(
+    TRITONBACKEND_Model* model, const char** location);
+
```

Comment:
Isn't [TRITONBACKEND_BackendArtifacts](https://github.com/triton-inference-server/core/blob/55358c65e222c559c72527b783abd5971dcd293b/include/triton/core/tritonbackend.h#L897-L915) essentially the same API? Then you can just search for `model.py` within the backend directory returned by this API.

Reply:
I don't think so, [TRITONBACKEND_BackendArtifacts](https://github.com/triton-inference-server/core/blob/55358c65e222c559c72527b783abd5971dcd293b/include/triton/core/tritonbackend.h#L897-L915) interact with a backend, which will be `python backend` and all related `.so` libs locations.

Under the hood of new python-based backends is `python backend`, so `backend` passed to  `TRITONBACKEND_BackendArtifacts(
    TRITONBACKEND_Backend* backend, TRITONBACKEND_ArtifactType* artifact_type,
    const char** location)` is associated with `***_python.so` and not with custom backend. Thus, this api won't work.. Let me know, if I am wrong

Reply:
Hm, I think based on current implementation details the `backend->Directory()` is being set to where python backend lives, ex: `/opt/tritonserver/backends/python`, so you're right it wouldn't work. 

In my head, I thought the backend would still be considered it's own "backend" and living in `/opt/tritonserver/backends/vllm` in the `TritonBackend` object for example, but under the hood just loaded `libtriton_python.so`. I think this would remove the need for storing "custom backends" and the need for adding a relatively duplicate new C API. But this would require some refactoring. @Tabrizian @tanmayv25 any thoughts here?

Reply:
You are correct, it will be living there, however we can't create a proper backend without `.so` file and `CreateBackend` can't be called on custom backends. I believe the idea was: we look for `.so` -> look for `model.py`. If we found `model.py` it still needs proper `libtriton_python.so` to load all handles from, so we create python backend and load `python`'s shared library first, then store path to  `/opt/tritonserver/backends/vllm/model.py` as a part of `TritonModel`.

Reply:
Is there a way to "hardcode" this special use of Python backend while somehow make `TRITONBACKEND_Artifacts`return the path to e.g. vllm backend?

I agree that having a separate API looks a bit awkward.

Reply:
`TRITONBACKEND_BackendArtifacts` would need somehow to return 2 locations then: to `libtriton_python.so` and vllm backend + this path to vllm backend should be know associated with the backend and not the model. One option can be to create a copy of python's backend for vllm backend ( to have an access to `.so` and `.py` stored in the same backend object). Still, I think even though we call these custom backends `backends`, they really just store model's executor, which should be associated with a Model object. Under the hood, it is still a `python backend`.

Reply:
We can introduce `TRITONBACKEND_ModelArtifacts` and make it more general.

Reply:
Although even with the copy, `TRITONBACKEND_BackendArtifacts` should return location to `.so` , and it will in most cases be in a different location than `model.py`

Reply:
Did we decide on how we will handle dependencies like vllm python package? For example, if we intended to go the route of execution environments or custom stubs, etc. - then I think we would **want** TRITONBACKEND_BackendArtifacts to return the path to the `/opt/tritonserver/backends/vllm` and have a custom stub in that directory. 

Would it make sense to copy the libtriton_python.so and stub into the `/vllm` backend directory, rather than routing it to the python backend dir, and keep them completely separate? @Tabrizian

Reply:
I think this would be very `vllm`-oriented. Should we then extend this idea (copying `.so` to the same custom backend folder) for every custom backend?

Reply:
I don't think it's that vllm oriented. Pytorch handler requires `pip install torch / torchvision` , Hugginface handler requires `pip install huggingface`. 

If we intended to handle these dependencies via stubs or environments, then I think we need to think about that. 

If we intend a somewhat simpler path for now and just install the dependencies in the container environment `pip install <backend>/requirements.txt` either baked into container, or on demand, then that's different.

Reply:
I see, thanks for clarification

Reply:
My original thought was going in the route of `pip install <backend>/requirements.txt` for first round of vllm container.